extensions/google-cloud-functions-http/runtime/src/main/java/io/quarkus/gcp/functions/http/QuarkusHttpFunction.java

```diff
@@ -0,0 +1,184 @@
+            }
+        } else {
+            errorWriter.println("Quarkus bootstrapped successfully.");
+            started = true;
+        }
+        deploymentStatus = error.toString();
+    }
+
+    @Override
+    public void service(HttpRequest request, HttpResponse response) throws IOException {
+        if (!started) {
+            throw new IOException(deploymentStatus);
+        }
+        dispatch(request, response);
+    }
+
+    private void dispatch(HttpRequest request, HttpResponse response) throws IOException {
+        try {
+            nettyDispatch(request, response);
+        } catch (IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+    }
+
+    private void nettyDispatch(HttpRequest request, HttpResponse response)
+            throws InterruptedException, IOException, ExecutionException {
+        String path = request.getPath();
+        Optional<String> host = request.getFirstHeader("Host");
+        DefaultHttpRequest nettyRequest = new DefaultHttpRequest(HttpVersion.HTTP_1_1,
+                HttpMethod.valueOf(request.getMethod()), path);
+        host.ifPresent(h -> nettyRequest.headers().set("Host", host));
+        for (Map.Entry<String, List<String>> header : request.getHeaders().entrySet()) {
+            nettyRequest.headers().add(header.getKey(), header.getValue());
+        }
+
+        HttpContent requestContent = LastHttpContent.EMPTY_LAST_CONTENT;
+        if (request.getContentLength() != 0) {
+            try (ByteArrayOutputStream buffer = new ByteArrayOutputStream()) {
+                int nRead;
+                byte[] data = new byte[BUFFER_SIZE];
+                while ((nRead = request.getInputStream().read(data, 0, data.length)) != -1) {
+                    buffer.write(data, 0, nRead);
+                }
+                buffer.flush();
+                byte[] byteArray = buffer.toByteArray();
+                ByteBuf body = Unpooled.wrappedBuffer(byteArray);
+                requestContent = new DefaultLastHttpContent(body);
+            }
+        }
+
+        ResponseHandler handler = new ResponseHandler(response);
+        VirtualClientConnection connection = VirtualClientConnection.connect(handler, VertxHttpRecorder.VIRTUAL_HTTP);
+        connection.sendMessage(nettyRequest);
+        connection.sendMessage(requestContent);
+        try {
+            handler.future.get();
+        } finally {
+            connection.close();
+        }
+    }
+
+    private class ResponseHandler implements VirtualResponseHandler {
+
+        ByteArrayOutputStream baos;
+        WritableByteChannel byteChannel;
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        final HttpResponse response;
+
+        public ResponseHandler(HttpResponse response) {
+            this.response = response;
+            this.baos = new ByteArrayOutputStream(BUFFER_SIZE);
```

Comment:
Yes, I saw that you did it for Azure Functions, I was not fan of it because it makes the code harder to read but after all it's important as not all response will have a body.
I updated the code with lazy init of tha baos.